Added initial support for services protected with Http Basic auth

sverkerEsriSE opened the pull request:
Since we are having lot of customers using OGC WMS services protected with Http Basic Auth, we need a way to use these in Javascript API applications.

Adding support for Basic Authentication to the resource-proxy seems like a doable way for now.

Much code was already there, but used for Windows Authentication. By adding another attribute, httpBasicAuth, to the xml config file and enabling PreAuthentication on the request, I was able to get this working.

bsvensson commented:
@sverkerEsriSE - the proxy.xsd file should also be updated for any changes to proxy configuration settings.

Did you test that this doesn't break existing workflow, e.g. with Windows Authentication?

Related issues: #157 and https://github.com/Esri/resource-proxy/issues/213#issuecomment-138792004

sverkerEsriSE commented:
@bsvensson - I have updated the .xsd file now.

I have been trying to test the Windows Authentication workflow, but this workflow seems to be a bit buggy (or at least hard to configure properly in both the proxy and on IIS).

I am testing against a Portal with Web Adaptor with Windows Authentication (only) and then a Web ApplicationBuilder app which tries to access the Portal. Since it detects a Portal, it will default to trying to get a token, even though Windows Authentication is (domain attribute) is set in the proxy. I think the flow in the code doesn't account for many possible scenarios.

For now, I will not have time to test possible scenarios, since I'll be going on leave for a year, starting tomorrow, but I cannot see why the workflow I added should break the existing workflow, as long as domain and httpBasicAuth are not both configured on the same serverUrl (or in that case, httpBasicAuth has to be false).

sverkerEsriSE commented:
On additonal note, there is also some configuration issues with the WMS services I am testing httpBasicAuth against, so three separate URL:s need to be configured. First the GetCapabilities request goes to a URL like http://<server>/<vendor>/wms, then additional requests (like GetMap) goes to http://<server>/<vendor>/ows and finally to http://<server>:80/<vendor>/ows

I think this is because how the server-side software is configured and its internal workings, but I can't seem to get the matchAll="false" to work correctly. As I figure, if matchAll="false" then a server URL of http://<server>/<vendor> should match all the above, maybe except for the one with :80 appended. Maybe the URL parser in the proxy should ignore the port of the protocol is HTTP and there is a :80 specified in the URL, or HTTPS and port 443.

I (or someone else) need to investigate this more, perhaps I am misunderstanding something.

nicogis commented:
This code has a problem:
WebRequest.PreAuthenticate property does: it’s a caching mechanism that caches the connection credentials for a given domain in the active process and resends it on subsequent requests. It does not send credentials on the first request but it will cache credentials on subsequent requests after authentication has succeeded.
So the idea is that you simply add the required Authorization header to the request on your own along with the authorization string that encodes the username and password 

```
string auth = "Basic " + Convert.ToBase64String(System.Text.Encoding.Default.GetBytes(user + ":" + pwd));
req.PreAuthenticate = true;
req.AuthenticationLevel = System.Net.Security.AuthenticationLevel.MutualAuthRequested;
req.Headers.Add("Authorization", auth);
```

afoxi911 commented:
was this function ever pushed into the latest release - i've tried both the 1.1.2 release and also the new .ashx file related to the changes made here and the schema change - no luck either way. need support for a basic auth esri service provided by gov client

mtrcn commented:
Any update about it?
